device/iot-device-samples/send-event/src/main/java/samples/com/microsoft/azure/sdk/iot/SendEvent.java

```diff
@@ -10,233 +10,79 @@
         }
 
-        System.out.println("Successfully created an IoT Hub client.");
-
-        // Set your token expiry time limit here
-        long time = 2400;
-        client.setOption("SetSASTokenExpiryTime", time);
-        System.out.println("Updated token expiry time to " + time);
-
-        client.registerConnectionStatusChangeCallback(new IotHubConnectionStatusChangeCallbackLogger(), new Object());
+        DeviceClient client = new DeviceClient(connString, IotHubClientProtocol.MQTT);
+        long clientSendInterval = 5; //Lower number here spawns send threads more frequently, can send more quickly. By default, value is 10
+        client.setOption("SetSendInterval", clientSendInterval);
+        countDownLatch = new CountDownLatch((int) numberOfMessagesToSend);
 
         client.open();
 
-        System.out.println("Opened connection to IoT Hub.");
-        System.out.println("Sending the following event messages:");
-
-        String deviceId = "MyJavaDevice";
-        double temperature = 0.0;
-        double humidity = 0.0;
-
-        for (int i = 0; i < numRequests; ++i)
+        final long startTime = System.currentTimeMillis();
+        for (int sentMessageCount = 0; sentMessageCount < numberOfMessagesToSend; sentMessageCount++)
         {
-            temperature = 20 + Math.random() * 10;
-            humidity = 30 + Math.random() * 20;
-
-            String msgStr = "{\"deviceId\":\"" + deviceId +"\",\"messageId\":" + i + ",\"temperature\":"+ temperature +",\"humidity\":"+ humidity +"}";
-
-            try
-            {
-                Message msg = new Message(msgStr);
-                msg.setContentType("application/json");
-                msg.setProperty("temperatureAlert", temperature > 28 ? "true" : "false");
-                msg.setMessageId(java.util.UUID.randomUUID().toString());
-                msg.setExpiryTime(D2C_MESSAGE_TIMEOUT);
-                System.out.println(msgStr);
-
-                EventCallback callback = new EventCallback();
-                client.sendEventAsync(msg, callback, msg);
-            }
-
-            catch (Exception e)
-            {
-                e.printStackTrace(); // Trace the exception
-            }
+            Message message = messagesToSend[sentMessageCount];
+            client.sendEventAsync(message, callbacks[sentMessageCount], message);
         }
 
-        System.out.println("Wait for " + D2C_MESSAGE_TIMEOUT / 1000 + " second(s) for response from the IoT Hub...");
+        //wait until all sent messages have been acknowledged by the iot hub, or until 90 minutes have passed
+        countDownLatch.await(90, TimeUnit.MINUTES);
 
-        // Wait for IoT Hub to respond.
-        try
-        {
-            Thread.sleep(D2C_MESSAGE_TIMEOUT);
-        }
+        final long stopTime = System.currentTimeMillis();
+        double secondsTaken = ((stopTime - startTime) / 1000.0);
+        System.out.println("Time taken: " + secondsTaken + " seconds");
```

Comment:
I still see time being calculated from start to stop. So we are still accounting for extra processing. Can we use the timestamps exactly at the time of message being sent and ack received to get the right perf numbers?

Reply:
Like I mentioned earlier, use the callback and stamp it when you get the ack. you should use context object to store this time for that message. Once you gather all the start and stop times for each message being sent you can average it by number of messages sent. This way of calculating time will be more accurate.

Reply:
sure, I can try that